ops/terraform/services/migrator/Jenkinsfile

```diff
@@ -0,0 +1,242 @@
+                return migratorStatus
+            }
+        }
+        sleep(heartbeatInterval)
+    }
+}
+
+// print formatted migrator messages
+void printMigratorMessage(message) {
+    body = message.body
+    println "Timestamp: ${java.time.LocalDateTime.now().toString()}"
+
+    if (body.stop_time == "n/a") {
+        println "Migrator ${body.pid} started at ${body.start_time} is running"
+    } else {
+        println "Migrator ${body.pid} started at ${body.start_time} is no longer running: '${body.code}' '${body.status}' as of ${body.stop_time}"
+    }
+}
+
+// purge all messages from the given
+def purgeQueue(String sqsQueueName, String region = 'us-east-1') {
+    sqsQueueUrl = getSqsQueueUrl(sqsQueueName, region)
+    result = sh(returnStdout: true, script: "aws sqs purge-queue --queue-url ${url} --region ${region}")
+    return result
+}
+
+// checks for indications of a running migrator deployment by looking for unconsumed SQS messages
+boolean canMigratorDeploymentProceed(String sqsQueueName, String region) {
+    println "Checking Migrator Queue ${sqsQueueName} State..."
+
+    if (sqsQueueExists(sqsQueueName)) {
+        sqsQueueUrl = getSqsQueueUrl(sqsQueueName)
+        println "Queue ${sqsQueueName} exists. Checking for messages in ${sqsQueueUrl} ..."
+        migratorMessages = receiveMigratorMessages(
+                url: sqsQueueUrl,
+                region: region,
+                maxMessages: 10,
+                visibilityTimeoutSeconds: 0,
+                waitTimeSeconds: 0)
+        if (migratorMessages?.isEmpty()) {
+            println "Queue ${sqsQueueName} is empty. Migrator deployment can proceed!"
+            return true
+        } else {
+            println "Queue ${sqsQueueName} has messages. Is there an old bfd-db-migrator instance running? Migrator deployment cannot proceed."
+            return false
+        }
+    } else {
+        println "Queue ${sqsQueueName} does not exist or is inaccessible. Migrator deployment can proceed!"
+        return true
+    }
+}
+
+// initializes, selects appropriate workspace, plans, and applies terraform
+def executeTerraform(Map args = [:]) {
+    amiId = args.amiId
+    bfdEnv = args.bfdEnv
+    gitBranchName = args.gitBranchName
+    createMigratorInstance = args.createMigratorInstance ?: false
+    heartbeatInterval = args.heartbeatInterval ?: 30
+
+    dir("${workspace}/ops/terraform/services/migrator") {
+        // Debug output terraform version
+        sh "terraform --version"
+
+        // Initilize terraform
+        sh "terraform init -no-color"
+
+        // - Attempt to create the desired workspace
+        // - Select the desired workspace
+        // NOTE: this is the terraform concept of workspace **NOT** Jenkins
+        sh """
+terraform workspace new "$bfdEnv" 2> /dev/null || true &&\
+terraform workspace select "$bfdEnv" -no-color
+"""
+        // Gathering terraform plan
+        echo "Timestamp: ${java.time.LocalDateTime.now().toString()}"
```

Comment:
any particular reason to use LocalDateTime here and Instant in the other place?

Reply:
Not really, no. Aspects of this were copy/pasta'd from the root Jenkinsfile. I'll make this more consistent.